File: go/ledger/ledger.go
// Package ledger provides double ledger accounting.
package ledger

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Entry represents one entry in the ledger.
type Entry struct {
	Date        string // "Y-m-d"
	Description string
	Change      int // in cents
}

// Ledger is a series of entries.
type Ledger []Entry

var localeConfig = map[string]struct {
	header                 []string
	dateFormat             string
	thousandsSeparator     string
	positiveCurrencyFormat string
	negativeCurrencyFormat string
}{
	"nl-NL": {
		header:                 []string{"Datum", "Omschrijving", "Verandering"},
		dateFormat:             "02-01-2006",
		thousandsSeparator:     ".",
		positiveCurrencyFormat: "%s %s,%02s ",
		negativeCurrencyFormat: "%s %s,%02s-",
	},
	"en-US": {
		header:                 []string{"Date", "Description", "Change"},
		dateFormat:             "01/02/2006",
		thousandsSeparator:     ",",
		positiveCurrencyFormat: "%s%s.%s ",
		negativeCurrencyFormat: "(%s%s.%s)",
	},
}

var currencies = map[string]string{
	"EUR": "€",
	"USD": "$",
}

// Less orders the ledger by date > description > change
func formatCurrency(locale, currency string, cents int) string {
	format := localeConfig[locale].positiveCurrencyFormat
	if cents < 0 {
		cents = -cents
		format = localeConfig[locale].negativeCurrencyFormat
	}
	centsStr := fmt.Sprintf("%02d", cents%100)
	parts := chunk(cents / 100)

	return fmt.Sprintf(format, currencies[currency], strings.Join(parts, localeConfig[locale].thousandsSeparator), centsStr)
}

func chunk(d int) []string {
	var parts []string
	num := strconv.Itoa(d)

	o := len(num) % 3
	if o == 0 {
		o = 3
	}

	for i := 0; i < len(num); i += o {
		if i != 0 {
			o = 3
		}
		parts = append(parts, num[i:i+o])
	}
	return parts
}

func truncate(s, cont string, l int) string {
	if len(s) > l {
		return s[:l-len(cont)] + cont
	}
	return s
}

func formatEntry(e Entry, locale, currency string) (string, error) {
	l := localeConfig[locale]
	t, err := time.Parse("2006-01-02", e.Date)
	if err != nil {
		return "", err
	}
	// Truncate long descriptions.
	description := truncate(e.Description, "...", 25)
	date := t.Format(l.dateFormat)
	amount := formatCurrency(locale, currency, e.Change)
	return fmt.Sprintf("%10s | %-25s | %13s\n", date, description, amount), nil
}

// FormatLedger formats a set of ledger entries.
func FormatLedger(currency string, locale string, entries []Entry) (string, error) {
	// Validation.
	if _, ok := currencies[currency]; !ok {
		return "", fmt.Errorf("invalid currency: %s", currency)
	}

	l, ok := localeConfig[locale]
	if !ok {
		return "", fmt.Errorf("unknown locale: %s", locale)
	}

	// Make a copy and sort it.
	var entriesCopy Ledger
	for _, e := range entries {
		entriesCopy = append(entriesCopy, e)
	}
	sort.Slice(entriesCopy, func(i, j int) bool {
		if entriesCopy[i].Date != entriesCopy[j].Date {
			return entriesCopy[i].Date < entriesCopy[j].Date
		}
		if entriesCopy[i].Description != entriesCopy[j].Description {
			return entriesCopy[i].Description < entriesCopy[j].Description
		}
		return entriesCopy[i].Change < entriesCopy[j].Change
	})

	ss := make([]string, len(entriesCopy))
	for i, et := range entriesCopy {
		s, err := formatEntry(et, locale, currency)
		if err != nil {
			return "", err
		}
		ss[i] = s
	}

	header := fmt.Sprintf("%-10s | %-25s | %s\n", l.header[0], l.header[1], l.header[2])
	return header + strings.Join(ss, ""), nil
}

File: python/ledger/ledger.py
# -*- coding: utf-8 -*-
"""Ledger formatting."""

# Changelog:
# * Import module, not classes.
# * Use dataclass for LedgerEntry.
# * Replace sorting block with sorted() and dataclass ordering.
# * Use string formating for header row.
# * Use date formatting for formatting the date.
# * Build rows as a list over a single string.
# * Move row formatting into LedgerEntry format().
# * Pull date formatting, description truncation out of conditional.
# * Fix description formatting to use a slice/ljust().
# * Use format string to format the row from parts.
# * Add a fmt_decimal that accepts ki and dec separators.
# * Move currency formatting into a function and use a template string.

import dataclasses
import datetime

# Currency symbols.
SYMBOL = {'USD': '$', 'EUR': '€'}


@dataclasses.dataclass(order=True)
class LedgerEntry:
    """Ledger entry."""
    date: datetime.datetime
    description: str
    change: int


@dataclasses.dataclass
class Formatter:
    """Formatter is used to format values."""

    # Date format string.
    date: str
    # Number separators: thousands and decimal.
    k_sep: str
    d_sep: str
    # Currency format string. (positive and negative amounts).
    currency: dict[bool, str]
    headers: list[str]

    def format_decimal(self, num: int) -> str:
        """Format a decimal number."""
        assert num >= 0, "Absolute values only"
        val = num // 100
        # Join thousands with the k_sep.
        whole = f"{val:_}".replace("_", self.k_sep)
        # Add the decimal part.
        return f'{whole}{self.d_sep}{num % 100:02}'

    def format_cur(self, amount: int, currency: str) -> str:
        """Format the currency change."""
        return self.currency[amount >= 0].format(
            symbol=SYMBOL[currency],
            num=self.format_decimal(abs(amount))
        )

    def format_date(self, date) -> str:
        """Format a date."""
        return date.strftime(self.date)

    @staticmethod
    def format_description(description) -> str:
        """Format a description."""
        if len(description) > 25:
            description = description[:22] + '...'
        return description

    def format_entry(self, entry: LedgerEntry, currency: str) -> str:
        """Format one entry."""
        date = self.format_date(entry.date)
        description = self.format_description(entry.description)
        change = self.format_cur(entry.change, currency)

        # Combine columns.
        return f'{date:<10} | {description:<25} | {change:>13}'

    def format(self, entries: list[LedgerEntry], currency: str) -> str:
        """Return entries formatted to a table."""
        parts = self.headers
        header = f'{parts[0]:<10} | {parts[1]:<25} | {parts[2]:<13}'
        rows = [header] + [self.format_entry(entry, currency) for entry in sorted(entries)]
        return '\n'.join(rows)


FMT = {
    'en_US': Formatter(
        date='%m/%d/%Y',
        k_sep = ',',
        d_sep = '.',
        currency = {True: '{symbol}{num} ', False: '({symbol}{num})'},
        headers = ['Date', 'Description', 'Change'],
    ),
    'nl_NL': Formatter(
        date='%d-%m-%Y',
        k_sep = '.',
        d_sep = ',',
        currency = {True: '{symbol} {num} ', False: '{symbol} -{num} '},
        headers = ['Datum', 'Omschrijving', 'Verandering'],
    ),
}


def create_entry(date: str, description: str, change: int) -> LedgerEntry:
    """Return a LedgerEntry."""
    pdate = datetime.datetime.strptime(date, '%Y-%m-%d')
    return LedgerEntry(pdate, description, change)


def format_entries(currency: str, locale: str, entries: list[LedgerEntry]) -> str:
    """Format LedgerEntry values."""
    return FMT[locale].format(entries, currency)
